### DISCUSSION TITLE
Multi-value query syntax discussion

### DISCUSSION THREAD

**Discussion by OxygenCobalt** [beetbox/beets#4792] [Ideas]
Continuation from #4743.

Current proposed query syntax:
- `foo | [...foo...]` = Match any instance of foo
- `[foo...]` = Match foo at the beginning
- `[...foo]` = Match foo at the end
- `[foo,bar,baz]` = Match exactly foo,bar,baz
- `[...foo,bar,baz...]` = Match foo,bar,baz in that order, but in any position
- `[...foo,bar,baz]` = Match foo,bar,baz in that order, but at the end
- `[foo,bar,baz...]` = Match foo,bar,baz in that order, but at the beginning

In-flux query syntax:
- `foo,bar,baz` = Match any instance of foo, bar, or baz, regardless of order

@jmbannon

**Comment by jmbannon** [comment]
A few out-loud thoughts:
- For most multi-tag read queries, I suspect people will not care about order
- For most multi-tag write queries, I suspect that these will be the most commonly used ops:
  - Explicitly set all values
  - Append one or many multi values
- Syntax should not collide with characters in artist's names, and be bash friendly

**Reply by jmbannon** [reply]
With that said, here's my proposed syntax:

Reading:
```
artists:{foo,bar,baz}  # contains at least one of foo, bar, baz
artists:[foo,bar,baz]  # contains all of foo, bar, baz (order does not matter), and nothing more
artists:[foo,bar,baz,*]  # contains all of foo, bar, baz, and allows extras
```

Writing:
```
artists:[*,boz]  # appends boz to the end
artists:[boz, *]  # Sets boz as the first element, shifts other elements over
artists:[biz, boz]  # Sets to biz, boz, overwriting everything else
```

For artists with the characters `{ } [ ] , *`, simply add `\` to escape it for the literal value

**Reply by jmbannon** [reply]
If order does matter, we can expand the syntax to be something like:

Reading:
```
artists:[[foo,bar,baz]]  # contains foo, bar, baz in that order
artists:[[*, foo, bar, baz]] # contains any number of values in the beginning, then foo, bar, baz in that order
artists:[[*, foo, bar, *, baz]]  # wildcard for zero or more values in those positions
```

**Reply by jmbannon** [reply]
Order queries could come later so we don't add too much in a single PR

**Reply by OxygenCobalt** [reply]
This looks good.

**Reply by jmbannon** [reply]
Another thought that should be discussed is, what should the actual values foo/bar/baz mean. As in, do we treat them as "must match this value exactly" or "contains this sequence of characters"? And also, what is the case sensitivity?

My opinion is "contains this sequence of chars" and is case insensitive for ease-of-use

**Reply by OxygenCobalt** [reply]
> My opinion is "contains this sequence of chars" and is case insensitive for ease-of-use

That would make sense, given that beets search is already extremely fuzzy.

**Reply by sampsyo** [reply]
It seems like it would be logical if all of these pieces was itself like a beets field query: namely, it would be a case-insensitive substring match by default, but could become a regex match with the `:` suffix, an exact match with `=`, and so on.

**Reply by jmbannon** [reply]
Maybe something like this?
```
artists:{foo,bar,baz}  # contains at least one of foo, bar, baz - substring
artists::{"fo[o]","b[a]r","^baz"}  # contains at least one of foo, bar, baz - supports regex
artists={foo,bar,baz}  # contains at least one of foo, bar, baz - exact match
```

The regex one might be tricky if we're using `[ ] { } ,` in the multi syntax

**Reply by sampsyo** [reply]
Absolutely. You could even imagine doing it attached to the individual pieces, if you want:

```
artists:{foo,:bar,=baz}
```

That would look for stuff where one artist contains `foo`, another artist matches the regex `bar`, and a third matches exactly `baz`. It's worth thinking about whether this is equivalent to writing:

```
artists:foo artists::bar artists:=baz
```

…that is, if the multi-query thing is shorthand for a conjunction of single-member matches.

**Reply by jmbannon** [reply]
`artists:{foo,:bar,=baz}` looks nice.

So one thing to remember is in the database, the multi value will look like `foo\␀bar\␀baz`, so `artists=baz` would not hold true. I imagine under the hood, we'd translate `{foo,:bar,=baz}` to a regex like `(.*foo.*|bar|(^|\␀)baz($|\␀))`

**Reply by sampsyo** [reply]
Right. Part of my suggestion there is that `artists:foo` behave differently when we know `artists` is a "list-valued" field, and that it would be equivalent to `artists:{foo}`. (We can do this because we know, when parsing the query, the types of all the fields.)

**Reply by jmbannon** [reply]
Ah okay, so basically implement the 'elementary operators' like `artists:foo` first, which only cares about a single element in the list (and knows how to query it properly via regex). Then, expand that by supporting the multi syntax, which really decomposes it into something like:

`artists:{foo,:bar,=baz}` = `artists:foo artists::bar artists:=baz`

**Reply by sampsyo** [reply]
That honestly sounds like a very good plan to me!

**Reply by jmbannon** [reply]
Sounds good to me as well! Will give better support to the single (`artists:foo`) query and simplify the contains_any (`artists:{foo}`) query implementation. The `artists:[ordered,list]` query will probably need its own special decomposition, as well as appending element(s) to the list.

While you're here @sampsyo , given @OxygenCobalt 's testimonial of everything working in the PR, mind giving it another pass and/or merge? :wink:

**Reply by sampsyo** [reply]
I am hoping to get to it very soon!
